examples/pwa.js

```diff
@@ -0,0 +1,321 @@
+    this.container = container;
+
+    win.AMP_SHADOW = true;
+    this.ampReadyPromise_ = new Promise(resolve => {
+      (window.AMP = window.AMP || []).push(resolve);
+    });
+    this.ampReadyPromise_.then(AMP => {
+      log('AMP LOADED:', AMP);
+    });
+
+    /** @private @const {string} */
+    this.baseUrl_ = null;
+    /** @private @const {?Element} */
+    this.host_ = null;
+    /** @private @const {?ShadowRoot} */
+    this.shadowRoot_ = null;
+    /** @private @const {!Array<string>} */
+    this.stylesheets_ = [];
+    /** @private @const {!Array<!Element>} */
+    this.scripts_ = [];
+    /** @private @const {...} */
+    this.viewer_ = null;
+  }
+
+  /**
+   */
+  clear() {
+    this.container.textContent = '';
+  }
+
+  /**
+   * @param {!Document} doc
+   * @param {string} url
+   */
+  show(doc, url) {
+    log('Show document:', doc, url);
+    this.container.textContent = '';
+
+    this.baseUrl_ = url;
+
+    this.host_ = this.win.document.createElement('div');
+    this.host_.classList.add('amp-doc-host');
+    this.container.appendChild(this.host_);
+
+    this.shadowRoot_ = this.host_.createShadowRoot();
+    log('Shadow root:', this.shadowRoot_);
+
+    this.ampReadyPromise_.then(AMP => {
+      const amp = AMP.attachShadowRoot(this.shadowRoot_);
+      this.viewer_ = amp.viewer;
+      this.viewer_.setMessageDeliverer(this.onMessage_.bind(this),
+          'http://localhost:8000');
+    });
+
+    // Head
+    log('head:', doc.head);
+    for (let n = doc.head.firstElementChild; n; n = n.nextElementSibling) {
+      const tagName = n.tagName;
+      const isMeta = tagName == 'META';
+      const isLink = tagName == 'LINK';
+      const name = n.getAttribute('name');
+      const rel = n.getAttribute('rel');
+      if (n.tagName == 'TITLE') {
+        this.title_ = n.textContent;
+        log('- title: ', this.title_);
+      } else if (isMeta && n.hasAttribute('charset')) {
+        // Ignore.
+      } else if (isMeta && name == 'viewport') {
+        // Ignore.
+      } else if (isLink && rel == 'canonical') {
+        this.canonicalUrl_ = n.getAttribute('href');
+        log('- canonical: ', this.canonicalUrl_);
+      } else if (isLink && rel == 'stylesheet') {
+        this.stylesheets_.push(n.getAttribute('href'));
+        log('- stylesheet: ', this.stylesheets_[this.stylesheets_.length - 1]);
+      } else if (n.tagName == 'STYLE') {
+        if (n.hasAttribute('amp-boilerplate')) {
+          // Ignore.
+          log('- ignored embedded style: ', n);
+        } else {
+          log('- embedded style: ', n);
+          this.shadowRoot_.appendChild(this.win.document.importNode(n, true));
+        }
+      } else if (n.tagName == 'SCRIPT') {
+        if (n.hasAttribute('src')) {
+          log('- src script: ', n);
+          this.scripts_.push(n);
+        } else {
+          log('- non-src script: ', n);
+          this.shadowRoot_.appendChild(this.win.document.importNode(n, true));
```

Comment:
So, JS scripts have to be shared across all AMP documents. In other words, we can register a custom element only once in the whole DOM document. However, non-JS scripts (application/json, application/json+ld) have to be in shadows.